### PR TITLE
Suggestion: Allow only patch updates on package.json

### LINKED ISSUE #11458: Allow patch updates only on package.json
## Test plan

https://github.com/brave/browser-laptop/issues/11458

----

### Description

Allow patch updates only on `package.json`.

https://github.com/brave/browser-laptop/pull/11453#issuecomment-335910450

> This change looks working: https://travis-ci.org/brave/browser-laptop/jobs/286654321#L3827, https://travis-ci.org/brave/browser-laptop/jobs/286672237#L3490
>
> It seems to me that that we would have to consider to start replacing carets with tildes on `package.json` for now to accept patch updates only. Minor updates could break builds easily and it will take quite long time (took me a day) to find which package causes that breakage. IMO we should eventually remove semver dependency and update `package.json` manually (until we get ready for using Greenkeeper #1701 for example ).

### Steps to Reproduce

  1. rm -rf node_modules && npm i
  3. Search `package.json` for carets `^`
  4. Replace `^` with tildes `~`
  2. `npm list`
  5. Replace the version numbers with ones on the list.

**Actual result:**

Minor updates sometimes causes a nasty bug.

**Expected result:**

Semver should be removed eventually. We could start by narrowing the version number range.

**Reproduces how often:** [What percentage of the time does it reproduce?]

### Brave Version

**about:brave info:**

**Reproducible on current live release:**

### PR DESCRIPTION
**Pull request by luixxiul**
This PR replaces carets with tildes to allow only patch updates on package.json.

Because of minor updates a nasty bug like #11454 and #10029 sometimes happens, which is hard to find what introduced them (there seems to be almost no way of detecting it automatically). The thing is that semver is just _a norm_ and developers can regard an update as a patch mistakenly, where it actually is a minor update. 

Forcing only patch updates is one step to remove semver, which introduces a security risk as well. IMHO, major / minor updates should be done via PR to merge the change after it is verified via Travis tests result that nothing should get corrupted due to that update, locally and remotely.

This PR is based on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16 which you should get with `npm list` on 7de16732862ed880ee46f96e00bc5cb8aa2c7f7d. I added comments to each update, referring to that gist, to make it easy to review.

As long as the packages.json is not manually updated, this PR should be valid.

Closes #11458

Auditors:

Test Plan:
1. npm install

Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).

Test Plan:

Reviewer Checklist:

Tests

- [ ] Adequate test coverage exists to prevent regressions
- [ ] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [ ] New files have MPL2 license header

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -84,111 +84,111 @@
   "homepage": "https://www.brave.com/",
   "dependencies": {
     "acorn": "3.2.0",
-    "ad-block": "^3.0.5",
+    "ad-block": "~3.0.5",
```
Comment (luixxiul) [inline package.json]: `ad-block@3.0.5` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L3

File: package.json [hunk 2]
```diff
@@ -84,111 +84,111 @@
...
     "aphrodite": "1.1.0",
-    "async": "^2.0.1",
-    "bat-balance": "^1.0.1",
-    "bat-client": "^1.0.5",
-    "bat-publisher": "^1.0.2",
-    "bignumber.js": "^4.0.4",
+    "async": "~2.5.0",
```
Comment (luixxiul) [inline package.json]: `async@2.5.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L58

File: package.json [hunk 3]
```diff
@@ -84,111 +84,111 @@
...
+    "bat-balance": "~1.0.1",
```
Comment (luixxiul) [inline package.json]: `bat-balance@1.0.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L325

File: package.json [hunk 4]
```diff
@@ -84,111 +84,111 @@
...
+    "bat-client": "~1.0.5",
```
Comment (luixxiul) [inline package.json]: `bat-client@1.0.5` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L375

File: package.json [hunk 5]
```diff
@@ -84,111 +84,111 @@
...
+    "bat-publisher": "~1.0.2",
```
Comment (luixxiul) [inline package.json]: `bat-publisher@1.0.2` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L772

File: package.json [hunk 6]
```diff
@@ -84,111 +84,111 @@
...
+    "bignumber.js": "~4.0.4",
```
Comment (luixxiul) [inline package.json]: `bignumber.js@4.0.4` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L794

File: package.json [hunk 7]
```diff
@@ -84,111 +84,111 @@
...
     "bloodhound-js": "brave/bloodhound",
-    "clipboard-copy": "^1.0.0",
-    "compare-versions": "^3.0.1",
-    "electron-localshortcut": "^0.6.0",
+    "clipboard-copy": "~1.2.0",
```
Comment (luixxiul) [inline package.json]: `clipboard-copy@1.2.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L827

File: package.json [hunk 8]
```diff
@@ -84,111 +84,111 @@
...
+    "compare-versions": "~3.1.0",
```
Comment (luixxiul) [inline package.json]: `compare-versions@3.1.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L827

File: package.json [hunk 9]
```diff
@@ -84,111 +84,111 @@
...
+    "electron-localshortcut": "~0.6.1",
```
Comment (luixxiul) [inline package.json]: `electron-localshortcut@0.6.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1386

File: package.json [hunk 10]
```diff
@@ -84,111 +84,111 @@
...
     "electron-squirrel-startup": "brave/electron-squirrel-startup",
-    "emoji-regex": "^6.5.1",
+    "emoji-regex": "~6.5.1",
```
Comment (luixxiul) [inline package.json]: `emoji-regex@6.5.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1457

File: package.json [hunk 11]
```diff
@@ -84,111 +84,111 @@
...
     "file-loader": "0.11.2",
-    "font-awesome": "^4.5.0",
+    "font-awesome": "~4.7.0",
```
Comment (luixxiul) [inline package.json]: `font-awesome@4.7.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1541

File: package.json [hunk 12]
```diff
@@ -84,111 +84,111 @@
...
     "font-awesome-webpack": "0.0.4",
-    "fs-extra": "^2.1.2",
-    "immutable": "^3.7.5",
-    "immutablediff": "^0.4.2",
+    "fs-extra": "~2.1.2",
```
Comment (luixxiul) [inline package.json]: `fs-extra@2.1.2` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1566

File: package.json [hunk 13]
```diff
@@ -84,111 +84,111 @@
...
+    "immutable": "~3.8.1",
```
Comment (luixxiul) [inline package.json]: `immutable@3.8.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1736

File: package.json [hunk 14]
```diff
@@ -84,111 +84,111 @@
...
+    "immutablediff": "~0.4.4",
```
Comment (luixxiul) [inline package.json]: `immutablediff@0.4.4` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1737

File: package.json [hunk 15]
```diff
@@ -84,111 +84,111 @@
...
     "immutablepatch": "brave/immutable-js-patch",
-    "l20n": "^3.5.1",
-    "lru-cache": "^1.0.0",
+    "l20n": "~3.5.1",
```
Comment (luixxiul) [inline package.json]: `l20n@3.5.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1849

File: package.json [hunk 16]
```diff
@@ -84,111 +84,111 @@
...
+    "lru-cache": "~1.1.1",
```
Comment (luixxiul) [inline package.json]: ` lru-cache@1.1.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1915

File: package.json [hunk 17]
```diff
@@ -84,111 +84,111 @@
...
     "moment": "~2.18.1",
-    "niceware": "^1.0.4",
-    "parse-torrent": "^5.8.1",
-    "prettier-bytes": "^1.0.3",
-    "prop-types": "^15.5.6",
-    "punycode": "^2.0.0",
+    "niceware": "~1.0.4",
```
Comment (luixxiul) [inline package.json]: `niceware@1.0.4` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1966

File: package.json [hunk 18]
```diff
@@ -84,111 +84,111 @@
...
+    "parse-torrent": "~5.8.3",
```
Comment (luixxiul) [inline package.json]: `parse-torrent@5.8.3` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2208

File: package.json [hunk 19]
```diff
@@ -84,111 +84,111 @@
...
+    "prettier-bytes": "~1.0.4",
```
Comment (luixxiul) [inline package.json]: `prettier-bytes@1.0.4` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2240

File: package.json [hunk 20]
```diff
@@ -84,111 +84,111 @@
...
+    "prop-types": "~15.6.0",
```
Comment (luixxiul) [inline package.json]: `prop-types@15.6.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2241

File: package.json [hunk 21]
```diff
@@ -84,111 +84,111 @@
...
+    "punycode": "~2.1.0",
```
Comment (luixxiul) [inline package.json]: `punycode@2.1.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2258

File: package.json [hunk 22]
```diff
@@ -84,111 +84,111 @@
...
     "qr-image": "3.2.0",
     "random-lib": "2.1.0",
-    "react": "~15.6.1",
-    "react-dnd": "^2.1.4",
-    "react-dnd-html5-backend": "^2.1.2",
-    "react-dom": "^15.5.4",
-    "string.prototype.endswith": "^0.2.0",
-    "string.prototype.startswith": "^0.2.0",
+    "react": "~15.6.2",
```
Comment (luixxiul) [inline package.json]: `react@15.6.2` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2266

File: package.json [hunk 23]
```diff
@@ -84,111 +84,111 @@
...
+    "react-dnd": "~2.5.1",
```
Comment (luixxiul) [inline package.json]: `react-dnd@2.5.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2279

File: package.json [hunk 24]
```diff
@@ -84,111 +84,111 @@
...
+    "react-dnd-html5-backend": "~2.5.1",
```
Comment (luixxiul) [inline package.json]: `react-dnd-html5-backend@2.5.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2294

File: package.json [hunk 25]
```diff
@@ -84,111 +84,111 @@
...
+    "react-dom": "~15.6.2",
```
Comment (luixxiul) [inline package.json]: `react-dom@15.6.2` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2296

File: package.json [hunk 26]
```diff
@@ -84,111 +84,111 @@
...
+    "string.prototype.endswith": "~0.2.0",
```
Comment (luixxiul) [inline package.json]: `string.prototype.endswith@0.2.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2730

File: package.json [hunk 27]
```diff
@@ -84,111 +84,111 @@
...
+    "string.prototype.startswith": "~0.2.0",
```
Comment (luixxiul) [inline package.json]: `string.prototype.startswith@0.2.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2731

File: package.json [hunk 28]
```diff
@@ -84,111 +84,111 @@
...
     "tablesort": "5.0.1",
     "tldjs": "1.6.2",
-    "tracking-protection": "1.1.x",
+    "tracking-protection": "~1.1.6",
```
Comment (luixxiul) [inline package.json]: `tracking-protection@1.1.6` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2738. To apply the same range specification.

File: package.json [hunk 29]
```diff
@@ -84,111 +84,111 @@
...
     "underscore": "1.8.3",
     "url-loader": "~0.6.2",
-    "webtorrent-remote": "^2.0.2"
+    "webtorrent-remote": "~2.0.2"
```
Comment (luixxiul) [inline package.json]: `webtorrent-remote@2.0.2` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L3678

File: package.json [hunk 30]
```diff
@@ -84,111 +84,111 @@
...
-    "babel-preset-env": "^1.1.6",
-    "babel-preset-react": "^6.1.18",
-    "babel-register": "^6.3.13",
-    "babel-runtime": "^6.3.13",
-    "base64-js": "^1.2.0",
-    "chai": "^3.4.1",
-    "chai-as-promised": "^5.1.0",
-    "co-mocha": "^1.1.2",
-    "cross-env": "^3.1.4",
+    "babel": "~6.23.0",
```
Comment (luixxiul) [inline package.json]: `babel@6.23.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L60

File: package.json [hunk 31]
```diff
@@ -84,111 +84,111 @@
...
+    "babel-core": "~6.26.0",
```
Comment (luixxiul) [inline package.json]: `babel-core@6.26.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L61

File: package.json [hunk 32]
```diff
@@ -84,111 +84,111 @@
...
+    "babel-loader": "~7.1.2",
```
Comment (luixxiul) [inline package.json]: `babel-loader@7.1.2` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L117

File: package.json [hunk 33]
```diff
@@ -84,111 +84,111 @@
...
+    "babel-plugin-transform-react-constant-elements": "~6.23.0",
```
Comment (luixxiul) [inline package.json]: `babel-plugin-transform-react-constant-elements@6.23.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L133

File: package.json [hunk 34]
```diff
@@ -84,111 +84,111 @@
...
+    "babel-plugin-transform-react-inline-elements": "~6.22.0",
```
Comment (luixxiul) [inline package.json]: `babel-plugin-transform-react-inline-elements@6.22.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L135

File: package.json [hunk 35]
```diff
@@ -84,111 +84,111 @@
...
+    "babel-polyfill": "~6.26.0",
```
Comment (luixxiul) [inline package.json]: `babel-polyfill@6.26.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L135

File: package.json [hunk 36]
```diff
@@ -84,111 +84,111 @@
...
+    "babel-preset-env": "~1.6.0",
```
Comment (luixxiul) [inline package.json]: `babel-preset-env@1.6.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L137

File: package.json [hunk 37]
```diff
@@ -84,111 +84,111 @@
...
+    "babel-preset-react": "~6.24.1",
```
Comment (luixxiul) [inline package.json]: `babel-preset-react@6.24.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L289

File: package.json [hunk 38]
```diff
@@ -84,111 +84,111 @@
...
+    "babel-register": "~6.26.0",
```
Comment (luixxiul) [inline package.json]: `babel-register@6.26.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L310

File: package.json [hunk 39]
```diff
@@ -84,111 +84,111 @@
...
+    "babel-runtime": "~6.26.0",
```
Comment (luixxiul) [inline package.json]: `babel-runtime@6.26.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L321

File: package.json [hunk 40]
```diff
@@ -84,111 +84,111 @@
...
+    "base64-js": "~1.2.1",
```
Comment (luixxiul) [inline package.json]: `base64-js@1.2.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L324

File: package.json [hunk 41]
```diff
@@ -84,111 +84,111 @@
...
+    "chai": "~3.5.0",
```
Comment (luixxiul) [inline package.json]: `chai@3.5.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L821

File: package.json [hunk 42]
```diff
@@ -84,111 +84,111 @@
...
+    "chai-as-promised": "~5.3.0",
```
Comment (luixxiul) [inline package.json]: `chai-as-promised@5.3.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L826

File: package.json [hunk 43]
```diff
@@ -84,111 +84,111 @@
...
+    "co-mocha": "~1.2.0",
```
Comment (luixxiul) [inline package.json]: `co-mocha@1.2.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L828

File: package.json [hunk 44]
```diff
@@ -84,111 +84,111 @@
...
+    "cross-env": "~3.2.4",
```
Comment (luixxiul) [inline package.json]: `cross-env@3.2.4` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L832

File: package.json [hunk 45]
```diff
@@ -84,111 +84,111 @@
...
     "css-loader": "~0.28.7",
-    "electron-builder": "^17.1.1",
-    "electron-chromedriver": "^1.7.1",
+    "electron-builder": "~17.10.0",
```
Comment (luixxiul) [inline package.json]: `electron-builder@17.10.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1029

File: package.json [hunk 46]
```diff
@@ -84,111 +84,111 @@
...
+    "electron-chromedriver": "~1.7.1",
```
Comment (luixxiul) [inline package.json]: `electron-chromedriver@1.7.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1276

File: package.json [hunk 47]
```diff
@@ -84,111 +84,111 @@
...
     "electron-packager": "brave/electron-packager",
     "electron-prebuilt": "brave/electron-prebuilt",
     "empty-port": "0.0.2",
-    "enzyme": "^2.9.1",
+    "enzyme": "~2.9.1",
```
Comment (luixxiul) [inline package.json]: `enzyme@2.9.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1459

File: package.json [hunk 48]
```diff
@@ -84,111 +84,111 @@
...
     "flow-bin": "~0.53.1",
-    "git-rev-sync": "^1.8.0",
-    "gulp": "^3.9.0",
-    "istanbul": "^1.1.0-alpha.1",
-    "joi": "^10.2.2",
+    "git-rev-sync": "~1.9.1",
```
Comment (luixxiul) [inline package.json]: `git-rev-sync@1.9.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1569

File: package.json [hunk 49]
```diff
@@ -84,111 +84,111 @@
...
+    "gulp": "~3.9.1",
```
Comment (luixxiul) [inline package.json]: `gulp@3.9.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1577

File: package.json [hunk 50]
```diff
@@ -84,111 +84,111 @@
...
+    "istanbul": "~1.1.0-alpha.1",
```
Comment (luixxiul) [inline package.json]: `istanbul@1.1.0-alpha.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1741

File: package.json [hunk 51]
```diff
@@ -84,111 +84,111 @@
...
+    "joi": "~10.6.0",
```
Comment (luixxiul) [inline package.json]: `joi@10.6.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1791

File: package.json [hunk 52]
```diff
@@ -84,111 +84,111 @@
...
-    "less": "^2.5.3",
-    "less-loader": "^2.2.1",
-    "level": "^1.7.0",
-    "mkdirp": "^0.5.1",
-    "mocha": "^2.3.4",
-    "mockery": "^2.1.0",
-    "muon-winstaller": "^2.5.5",
-    "ncp": "^2.0.0",
-    "node-gyp": "^3.3.1",
+    "json-loader": "~0.5.7",
```
Comment (luixxiul) [inline package.json]: `json-loader@0.5.7` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1846

File: package.json [hunk 53]
```diff
@@ -84,111 +84,111 @@
...
+    "jsonfile": "~2.4.0",
```
Comment (luixxiul) [inline package.json]: `jsonfile@2.4.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1847

File: package.json [hunk 54]
```diff
@@ -84,111 +84,111 @@
...
+    "less": "~2.7.2",
```
Comment (luixxiul) [inline package.json]: `less@2.7.2` to https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1850

File: package.json [hunk 55]
```diff
@@ -84,111 +84,111 @@
...
+    "less-loader": "~2.2.3",
```
Comment (luixxiul) [inline package.json]: `less-loader@2.2.3` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1861

File: package.json [hunk 56]
```diff
@@ -84,111 +84,111 @@
...
+    "level": "~1.7.0",
```
Comment (luixxiul) [inline package.json]: `level@1.7.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1867

File: package.json [hunk 57]
```diff
@@ -84,111 +84,111 @@
...
+    "mkdirp": "~0.5.1",
```
Comment (luixxiul) [inline package.json]: `mkdirp@0.5.1` to https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1916

File: package.json [hunk 58]
```diff
@@ -84,111 +84,111 @@
...
+    "mocha": "~2.5.3",
```
Comment (luixxiul) [inline package.json]: `mocha@2.5.3` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1918

File: package.json [hunk 59]
```diff
@@ -84,111 +84,111 @@
...
+    "mockery": "~2.1.0",
```
Comment (luixxiul) [inline package.json]: `mockery@2.1.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1936

File: package.json [hunk 60]
```diff
@@ -84,111 +84,111 @@
...
+    "muon-winstaller": "~2.5.5",
```
Comment (luixxiul) [inline package.json]: `muon-winstaller@2.5.5` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1938

File: package.json [hunk 61]
```diff
@@ -84,111 +84,111 @@
...
+    "ncp": "~2.0.0",
```
Comment (luixxiul) [inline package.json]: `ncp@2.0.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1965

File: package.json [hunk 62]
```diff
@@ -84,111 +84,111 @@
...
+    "node-gyp": "~3.6.2",
```
Comment (luixxiul) [inline package.json]: `node-gyp@3.6.2` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L1970

File: package.json [hunk 63]
```diff
@@ -84,111 +84,111 @@
...
     "node-libs-browser": "~2.0.0",
-    "node-static": "^0.7.7",
-    "nsp": "^2.2.0",
+    "node-static": "~0.7.10",
```
Comment (luixxiul) [inline package.json]: `node-static@0.7.10` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2152

File: package.json [hunk 64]
```diff
@@ -84,111 +84,111 @@
...
+    "nsp": "~2.8.0",
```
Comment (luixxiul) [inline package.json]: `nsp@2.8.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2158

File: package.json [hunk 65]
```diff
@@ -84,111 +84,111 @@
...
     "pre-commit": "brave/pre-commit",
-    "pre-push": "^0.1.1",
-    "react-addons-perf": "^15.2.1",
-    "react-addons-test-utils": "^15.4.1",
-    "react-test-renderer": "^15.5.4",
-    "request": "^2.81.0",
-    "sinon": "^1.17.6",
+    "pre-push": "~0.1.1",
```
Comment (luixxiul) [inline package.json]: `pre-push@0.1.1` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2238

File: package.json [hunk 66]
```diff
@@ -84,111 +84,111 @@
...
+    "react-addons-perf": "~15.4.2",
```
Comment (luixxiul) [inline package.json]: `react-addons-perf@15.4.2` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2275

File: package.json [hunk 67]
```diff
@@ -84,111 +84,111 @@
...
+    "react-addons-test-utils": "~15.6.2",
```
Comment (luixxiul) [inline package.json]: `react-addons-test-utils@15.6.2` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2278

File: package.json [hunk 68]
```diff
@@ -84,111 +84,111 @@
...
+    "react-test-renderer": "~15.6.2",
```
Comment (luixxiul) [inline package.json]: `react-test-renderer@15.6.2` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2301

File: package.json [hunk 69]
```diff
@@ -84,111 +84,111 @@
...
+    "request": "~2.82.0",
```
Comment (luixxiul) [inline package.json]: `request@2.82.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2304

File: package.json [hunk 70]
```diff
@@ -84,111 +84,111 @@
...
+    "sinon": "~1.17.7",
```
Comment (luixxiul) [inline package.json]: `sinon@1.17.7` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2365

File: package.json [hunk 71]
```diff
@@ -84,111 +84,111 @@
...
     "spectron": "brave/spectron#chromium60",
     "standard": "9.0.0",
     "style-loader": "~0.18.2",
-    "uglifyjs-webpack-plugin": "^1.0.0-beta.2",
-    "uuid": "^3.0.1",
+    "uglifyjs-webpack-plugin": "~1.0.0-beta.2",
```
Comment (luixxiul) [inline package.json]: `uglifyjs-webpack-plugin@1.0.0-beta.2` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2743

File: package.json [hunk 72]
```diff
@@ -84,111 +84,111 @@
...
+    "uuid": "~3.1.0",
```
Comment (luixxiul) [inline package.json]: `uuid@3.1.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L2873

File: package.json [hunk 73]
```diff
@@ -84,111 +84,111 @@
...
     "webdriverio": "4.7.1",
     "webpack": "~3.4.1",
     "webpack-dev-server": "~2.6.1",
-    "webpack-notifier": "^1.2.1",
-    "xml2js": "^0.4.15"
+    "webpack-notifier": "~1.5.0",
```
Comment (luixxiul) [inline package.json]: `webpack-notifier@1.5.0` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L3646

File: package.json [hunk 74]
```diff
@@ -84,111 +84,111 @@
...
+    "xml2js": "~0.4.19"
```
Comment (luixxiul) [inline package.json]: `xml2js@0.4.19` on https://gist.github.com/luixxiul/1cf8689da05e509d45176588730e5a16#file-npm-txt-L3994

**Comment by petemill** [comment]
If we do this, then we could use an automated service like https://greenkeeper.io/ which monitors npm dependencies, then sends PRs when each npm dependency has a new minor / major version, allowing our tests to run with only that update

**Comment by evq** [comment]
I'm not sure this will have much of an effect, we have a `package-lock.json` and my understanding is that this actually freezes to an exact version including patch. So it should be possible to determine the source (commit) that introduced any breaking package upgrade.

I think `package-lock.json` changes can be quite hard to read manually, so maybe we need to look for some automation so changes can be reviewed in a concise format?

**Comment by luixxiul** [comment]
My intension here is to prevent minor updates (which actually the author should name as major ones) from breaking the browser. On the 1st post:

> Because of minor updates a nasty bug like #11454 and #10029 sometimes happens, which is hard to find what introduced them (there seems to be almost no way of detecting it automatically). The thing is that semver is just a norm and developers can regard an update as a patch mistakenly, where it actually is a minor update.

**Comment by luixxiul** [comment]
(patch updates may break the stuff as well, so ultimately removing the semver at all can be the better solution here)

**Comment by luixxiul** [comment]
> my understanding is that this actually freezes to an exact version including patch.

If it is correct, why wouldn't we consider to remove the semver at all? :-)

**Comment by evq** [comment]
oh wow. I looked into this a little more (thanks for the stackoverflow link, found this via a comment) and it turns out there is an NPM issue discussing a change around `5.3.0` where `package-lock.json` wasn't pinning an exact version. A comment on that issue seems to indicate that the latest version has the behavior one would expect, namely:

>If you have a package.json and you run npm i we generate a package-lock.json from it.

>If you run npm i against that package.json and package-lock.json, the latter will never be updated, even if the package.json would be happy with newer versions.

>If you manually edit your package.json to have different ranges and run npm i and those ranges aren't compatible with your package-lock.json then the latter will be updated with version that are compatible with your package.json. Further runs of npm i will be as with 2 above.

 https://github.com/npm/npm/issues/17979#issuecomment-332701215

Perhaps this led to the issues you mentioned?

I just tested a clean `npm install` with npm `5.5.1`, and there were no version changes to `package-lock.json`. We could set the engine version of npm in `package.json` to `5.5.1` or another newer version.

**Comment by luixxiul** [comment]
Closing for now to see if another issue would happen after upgrading npm version.

**Pull request closed without merging**